File: semantica/merge.py
import time
import tracemalloc

def merge_sort(arr):
    if len(arr) <= 1:
        return arr.copy()

    mid = len(arr) // 2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])

    return merge(left, right)

def merge(left, right):
    if not left:
        return right.copy()
    if not right:
        return left.copy()

    if left[0] <= right[0]:
        return [left[0]] + merge(left[1:], right)
    else:
        return [right[0]] + merge(left, right[1:])

if __name__ == "__main__":
    with open("numeros.txt") as f:
        line = f.read().strip()
        arr = list(map(int, line.split(",")))

    tracemalloc.start()
    start = time.perf_counter()

    merge_sort(arr)

    end = time.perf_counter()
    current, peak = tracemalloc.get_traced_memory()
    tracemalloc.stop()

    # print(result)
    print(f"\nTiempo de ejecución del algoritmo: {(end - start)*1000:.3f} ms")
    print(f"Memoria actual usada: {current / 1024:.2f} KB")
    print(f"Memoria pico usada: {peak / 1024:.2f} KB")

File: semantica/merge.go
package main

import (
    "fmt"
    "io/ioutil"
    "strconv"
    "strings"
    "time"
    "runtime"
)

func mergeSort(arr []int) []int {
    if len(arr) <= 1 {
        return arr
    }

    queue := make([][]int, len(arr))
    for i, x := range arr {
        queue[i] = []int{x}
    }

    for len(queue) > 1 {
        left := queue[0]
        right := queue[1]
        queue = append(queue[2:], merge(left, right))
    }

    return queue[0]
}

func merge(left, right []int) []int {
    result := make([]int, 0, len(left)+len(right))
    i, j := 0, 0

    for i < len(left) && j < len(right) {
        if left[i] <= right[j] {
            result = append(result, left[i])
            i++
        } else {
            result = append(result, right[j])
            j++
        }
    }

    result = append(result, left[i:]...)
    result = append(result, right[j:]...)
    return result
}

func main() {
    data, err := ioutil.ReadFile("numeros.txt")
    if err != nil {
        panic(err)
    }

    line := strings.TrimSpace(string(data))
    parts := strings.Split(line, ",")

    var arr []int
    for _, part := range parts {
        n, err := strconv.Atoi(part)
        if err != nil {
            panic(err)
        }
        arr = append(arr, n)
    }

    var m1, m2 runtime.MemStats
    runtime.ReadMemStats(&m1)
    start := time.Now()

    mergeSort(arr)

    duration := time.Since(start)
    runtime.ReadMemStats(&m2)

    usedAlloc := m2.Alloc - m1.Alloc
    totalAlloc := m2.TotalAlloc - m1.TotalAlloc

    // fmt.Println(result)

    fmt.Printf("\nTiempo de ejecución del algoritmo: %.3f ms\n", float64(duration.Microseconds())/1000)
    fmt.Printf("Memoria actual usada (Alloc): %d bytes (%.2f KB)\n", usedAlloc, float64(usedAlloc)/1024)
    fmt.Printf("Memoria total asignada (TotalAlloc): %d bytes (%.2f KB)\n", totalAlloc, float64(totalAlloc)/1024)
}
